map_model/src/make/turns.rs

```diff
@@ -387,16 +337,58 @@ fn remove_merging_turns(map: &Map, input: Vec<Turn>, turn_type: TurnType) -> Vec
             continue;
         }
 
-        // From one to many is fine
-        if group.iter().map(|t| t.id.src).collect::<HashSet<_>>().len() == 1 {
+        let num_src_lanes = group.iter().map(|t| t.id.src).collect::<HashSet<_>>().len();
+        let num_dst_lanes = group.iter().map(|t| t.id.dst).collect::<HashSet<_>>().len();
+
+        // Allow all turns from one to many
+        if num_src_lanes == 1 {
             turns.extend(group);
             continue;
         }
 
-        // We have multiple lanes all with a turn to the same destination road. Most likely, only
-        // the rightmost or leftmost can actually make the turn.
-        // TODO If OSM turn restrictions explicitly have something like "left|left|", then there
-        // are multiple source lanes!
+        // If the number of source and destination lanes is the same, match them up in order,
+        // without any crossing.
+        if num_src_lanes == num_dst_lanes {
+            // But we want to match things up -- leftmost turn lane leads to leftmost destination.
+            let mut src_lanes_in_order: Vec<LaneID> = group.iter().map(|t| t.id.src).collect();
+            src_lanes_in_order.sort_by_key(|l| map.get_parent(*l).dir_and_offset(*l).1);
+            let mut dst_lanes_in_order: Vec<LaneID> = group.iter().map(|t| t.id.dst).collect();
+            dst_lanes_in_order.sort_by_key(|l| map.get_parent(*l).dir_and_offset(*l).1);
+
+            for t in group {
+                let src_order = src_lanes_in_order
+                    .iter()
+                    .position(|l| t.id.src == *l)
+                    .unwrap();
+                let dst_order = dst_lanes_in_order
+                    .iter()
+                    .position(|l| t.id.dst == *l)
+                    .unwrap();
+                if src_order == dst_order {
+                    turns.push(t);
+                }
+            }
+            continue;
+        }
+
+        // If src < dst and src isn't 1, then one source lane gets to access multiple destination
+        // lanes. For now, just give up figuring this out, and allow all combinations.
+        //
+        // TODO https://wiki.openstreetmap.org/wiki/Relation:connectivity may have hints about a
+        // better algorithm.
+        if num_src_lanes < num_dst_lanes {
```

Comment:
An example of this would be like 3 left turn lanes leading to 2 possible lanes. Where's the merging happen?

Reply:
The case when these are equal is handled above: `// If the number of source and destination lanes is the same, match them up in order, without any crossing` and it does require extra processing to line up the leftmost lanes, second-to-leftmost lanes, etc.

And the case below I think is also necessary, unless we come up with better rules for how to make 3 sources merge into 2 destinations.

Reply:
ah, right, I didn't realize this was effectively an "else if"